compiler/internal/codegen/testdata/TestCodeGenMain__variants.golden

```diff
@@ -1505,10 +1517,145 @@ func DI(ctx context.Context) error {
+		switch field {
+		case "DefaultFooParams":
+			obj.DefaultFooParams = config.CreateValue[FooParams](EncoreInternal_UnmarshalConfig_FooParams(itr, append(path, "DefaultFooParams")), append(path, "DefaultFooParams"))
+		case "ReadOnly":
+			obj.ReadOnly = config.CreateValue[bool](itr.ReadBool(), append(path, "ReadOnly"))
+		case "MagicNumber":
+			obj.MagicNumber = itr.ReadInt()
+		case "Sub":
+			obj.Sub = EncoreInternal_UnmarshalConfig_SubType(itr, append(path, "Sub"))
+		}
+		return true
+	})
+	return
+}
+
+func EncoreInternal_UnmarshalConfig_SubType(itr *jsoniter.Iterator, path []string) (obj SubType) {
+	itr.ReadObjectCB(func(itr *jsoniter.Iterator, field string) bool {
+		switch field {
+		case "FieldA":
+			obj.FieldA = config.CreateValue[string](itr.ReadString(), append(path, "FieldA"))
+		case "FieldB":
+			obj.FieldB = config.CreateValue[[]string](config.ReadArray[string](itr, func(itr *jsoniter.Iterator, idx int) string {
+				return itr.ReadString()
+			}), append(path, "FieldB"))
+		case "FieldC":
+			obj.FieldC = config.ReadArray[int](itr, func(itr *jsoniter.Iterator, idx int) int {
+				return config.CreateValue[int](itr.ReadInt(), append(path, strconv.Itoa(idx)))
+			})
+		case "FieldD":
+			obj.FieldD = config.ReadMap[int, string](itr, func(itr *jsoniter.Iterator, keyAsString string) (int, string) {
+				// Decode the map key from the JSON string to the underlying type it needs to be
+				keyDecoder := &etype.Marshaller{}
+				key := keyDecoder.ToInt("keyAsString", key, true)
+				if keyDecoder.LastError != nil {
+					panic(fmt.Sprintf("unable to decode the config: %v", keyDecoder.LastError))
+				}
+
+				return key, config.CreateValue[string](itr.ReadString(), append(path, keyAsString))
+			})
+		case "FieldE":
+			obj.FieldE = func() (obj struct {
+				SubA string
+				SubB FooParams
+				Blah time.Time
+				Foo  []byte
+				Json json.RawMessage
+			}) {
+				itr.ReadObjectCB(func(itr *jsoniter.Iterator, field string) bool {
+					switch field {
+					case "SubA":
+						obj.SubA = config.CreateValue[string](itr.ReadString(), append(path, "SubA"))
+					case "SubB":
+						obj.SubB = EncoreInternal_UnmarshalConfig_FooParams(itr, append(path, "SubB"))
+					case "Blah":
+						obj.Blah = func() (rtn time.Time) {
+							decoder := &etype.Marshaller{}
+
+							rtn = decoder.ToTime("value", itr.ReadString(), true)
+
+							if decoder.LastError != nil {
+								panic(fmt.Sprintf("unable to decode the config: %v", decoder.LastError))
+							}
+
+							return
+						}()
+					case "Foo":
+						obj.Foo = func() (rtn []byte) {
+							decoder := &etype.Marshaller{}
+
+							rtn = decoder.ToBytes("value", itr.ReadString(), true)
+
+							if decoder.LastError != nil {
+								panic(fmt.Sprintf("unable to decode the config: %v", decoder.LastError))
+							}
+
+							return
+						}()
+					case "Json":
+						obj.Json = func() (rtn json.RawMessage) {
+							decoder := &etype.Marshaller{}
```

Comment:
why not use a shared marshaller?

Reply:
Because error tracking needs to occur within the single read - and the `readType` code generator is not aware it's been called within another call to `readType` where a shared marshaller could be initialised.

Reply:
Not part of this exercise, but I guess the decoder methods could easily be turned into static helpers. would reduce the generated code and overall complexity.